File: longest_substring_without_repeating_char.cpp
#include "header.h"

int lengthOfLongestSubstring(string s) {
    int res = 0;
    if(s.size()<2) return s.size();
    int sz = s.size();
    int l = 0, r = 0;
    unordered_set<char> set;
    while(r<sz){
        if(set.find(s[r]) == set.end()){
            set.insert(s[r]);
            res = std::max(res,r-l+1);
            r++;
            continue;
        }
        do{
            set.erase(s[l]);
        }while(s[l++] != s[r]);
    }
    return res;
}

int main(){
    printf("%d\n", lengthOfLongestSubstring("abcabcbb")); //abc
    printf("%d\n", lengthOfLongestSubstring("bbbbb")); //b
    return 0;
}

File: longest_substring_without_repeating_char.py
"""
Given a string, find the length of the longest substring without repeating characters. For example, the longest substring without repeating letters for "abcabcbb" is "abc", which the length is 3. For "bbbbb" the longest substring is "b", with the length of 1.
"""

class Solution:
    # @return an integer
    def lengthOfLongestSubstring(self, s):
        m = [0 for i in range(256)]
        l,r=0,0
        res = 0
        while r<len(s):
            c = ord(s[r])
            if m[c] == 0:
                m[c] += 1
                r += 1
                res = max(r-l, res)
            else:
                while True:
                    m[ord(s[l])]-=1
                    l+=1
                    if s[l-1] == s[r]:
                        break
        return res


#print Solution().lengthOfLongestSubstring("abcabcbb")
print Solution().lengthOfLongestSubstring("wlrbbmqbhcdarzowkkyhiddqscdxrjmowfrxsjybldbefsarcbynecdyggxxpklorellnmpapqfwkhopkmco")
